### PR TITLE
Fail fast + docs around mixing major Scala versions serializing

### PR DESCRIPTION
**Pull request by johanandren**
Refs #22004

### CODE REVIEW

**Comment by ktoso** [comment]
Also add a note do the binary compatibility page, about serialization compatibility across scala versions

File: akka-actor/src/main/scala/akka/actor/Actor.scala [hunk 1]
```diff
@@ -73,6 +73,9 @@ final case class Identify(messageId: Any) extends AutoReceivedMessage with NotIn
  */
 @SerialVersionUID(1L)
 final case class ActorIdentity(correlationId: Any, ref: Option[ActorRef]) {
+  require(ref.isEmpty || ref.get != null, "ActorIdentity created with ref = Some(null) is not allowed, " +
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/Actor.scala]: I'd prefer `if` here to avoid allocating the function

**Review by patriknw — APPROVED** [review]
LGTM, with a small adjustment

**Review by drewhk — APPROVED** [review]

File: akka-actor/src/main/scala/akka/actor/Actor.scala [hunk 2]
```diff
@@ -73,6 +73,11 @@ final case class Identify(messageId: Any) extends AutoReceivedMessage with NotIn
  */
 @SerialVersionUID(1L)
 final case class ActorIdentity(correlationId: Any, ref: Option[ActorRef]) {
+  if (ref.isEmpty || ref.get != null) {
```
Comment (johanandren) [inline akka-actor/src/main/scala/akka/actor/Actor.scala]: D'oh, forgot to invert when I replaced `require`

File: akka-docs/rst/common/binary-compatibility-rules.rst [hunk 3]
```diff
@@ -151,3 +151,12 @@ called ``MiMa`` for short, for enforcing binary compatibility is kept where it w
...
 manual exception overrides may be put in place if the change happened to be in an Internal API for example.
+
+Serialization compatibility across Scala versions
+=================================================
+
+Scala does not maintain serialization compatibility across major versions. This means that if Java serialization is used
+there is no guarantee objects can be cleanly deserialized if serialized with a different version of Scala.
+
+The internal Akka Protobuf serializers that can be enabled with ``enable-additional-serialization-bindings``
+does not suffer from this problem.
```
Comment (ktoso) [inline akka-docs/rst/common/binary-compatibility-rules.rst]:
```
or even the ``akka.actor.allow-java-serialization = off``, which is preferable from a security stand point.

```

**Pull request merged**